File: HSTracker/UIs/ImageCache.m
/*
 * This file is part of the HSTracker package.
 * (c) Benjamin Michotte <bmichotte@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Created on 16/02/16.
 */
#import "ImageCache.h"
#import "Card.h"

typedef NS_ENUM(NSInteger, FromDestination)
{
    FromDestination_Bundle,
    FromDestination_Assets,
    FromDestination_Path,
};

@implementation ImageCache

+ (NSImage *)frameImageMask
{
  return [self imageNamed:@"frame_mask.png" from:FromDestination_Assets];
}

+ (NSImage *)smallCardImage:(Card *)card
{
  NSString *image = [[[card.englishName lowercaseString] replace:RX(@"[ ']")
                                                            with:@"-"] replace:RX(@"[:.!]")
                                                                          with:@""];
  return [self imageNamed:[NSString stringWithFormat:@"%@.png", image] from:FromDestination_Bundle];
}

+ (NSImage *)gemImage:(NSString *)rarity
{
  NSString *image;
  if ([rarity isEqualToString:@"free"]) {image = @"gem_rarity_free";}
  else if ([rarity isEqualToString:@"common"]) {image = @"gem_rarity_common";}
  else if ([rarity isEqualToString:@"rare"]) {image = @"gem_rarity_rare";}
  else if ([rarity isEqualToString:@"epic"]) {image = @"gem_rarity_epic";}
  else if ([rarity isEqualToString:@"legendary"]) {image = @"gem_rarity_legendary";}
  else {return nil;}

  return [self imageNamed:image from:FromDestination_Assets];
}

+ (NSImage *)frameDeckImage
{
  return [self imageNamed:@"frame_deck" from:FromDestination_Assets];
}

+ (NSImage *)frameImage:(NSString *)rarity
{
  NSString *image;
  if ([rarity isEqualToString:@"common"]) {image = @"frame_rarity_common";}
  else if ([rarity isEqualToString:@"rare"]) {image = @"frame_rarity_rare";}
  else if ([rarity isEqualToString:@"epic"]) {image = @"frame_rarity_epic";}
  else if ([rarity isEqualToString:@"legendary"]) {image = @"frame_rarity_legendary";}
  else {image = @"frame";}

  return [self imageNamed:image from:FromDestination_Assets];
}

+ (NSImage *)frameLegendary
{
  return [self imageNamed:@"frame_legendary" from:FromDestination_Assets];
}

+ (NSImage *)frameCount:(NSNumber *)number
{
  return [self imageNamed:[NSString stringWithFormat:@"frame_%@", number] from:FromDestination_Assets];
}

+ (NSImage *)frameCountbox
{
  return [self imageNamed:@"frame_countbox" from:FromDestination_Assets];
}

+ (NSImage *)frameCountboxDeck
{
  return [self imageNamed:@"frame_countbox_deck" from:FromDestination_Assets];
}

+ (NSImage *)imageNamed:(NSString *)path from:(FromDestination)from
{
  switch (from) {
    case FromDestination_Bundle: {
      NSString *fullPath = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:path];
      DDLogVerbose(@"Opening image %@", fullPath);
      return [[NSImage alloc] initWithContentsOfFile:fullPath];
    }
    case FromDestination_Assets: {
      DDLogVerbose(@"Opening image %@", path);
      return [NSImage imageNamed:path];
    }
    case FromDestination_Path: {
      DDLogVerbose(@"Opening image %@", path);
      return [[NSImage alloc] initWithContentsOfFile:path];
    }
    default:
      return nil;
  }
}
@end

File: HSTracker/UIs/ImageCache.swift
/*
 * This file is part of the HSTracker package.
 * (c) Benjamin Michotte <bmichotte@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Created on 16/02/16.
 */

import Foundation

enum FromDestination: Int {
    case Bundle,
        Assets,
        Path
}

struct ImageCache {

    static func cardImage(card: Card) -> NSImage? {
        if let appSupport = NSSearchPathForDirectoriesInDomains(
            .ApplicationSupportDirectory,
            .UserDomainMask, true).first {

            let path = "\(appSupport)/HSTracker/cards/\(card.id).png"
            let image = imageNamed(path, from: .Path)
            return image
        }
        return nil
    }

    static func smallCardImage(card: Card) -> NSImage? {
        return imageNamed("\(card.id).png", from: .Bundle)
    }

    static func gemImage(rarity: Rarity?) -> NSImage? {
        return image("gem", rarity: rarity)
    }

    static func frameImage(rarity: Rarity?) -> NSImage? {
        return image("frame", rarity: rarity)
    }

    private static func image(base: String, rarity: Rarity?) -> NSImage? {
        var image: String = "\(base)"
        if let rarity = rarity {
            switch rarity {
            case .Common: image += "_common"
            case .Rare: image += "_rare"
            case .Epic: image += "_epic"
            case .Legendary: image += "_legendary"
            case .Golden: image += "_golden"
            default: break
            }
        }
        return imageNamed(image, from: .Assets)
    }

    static func darkenImage() -> NSImage? {
        return imageNamed("dark", from: .Assets)
    }

    static func fadeImage() -> NSImage? {
        return imageNamed("fade", from: .Assets)
    }

    static func frameLegendary() -> NSImage? {
        return imageNamed("icon_legendary", from: .Assets)
    }

    static func frameCountbox(rarity: Rarity?) -> NSImage? {
        return image("countbox", rarity: rarity)
    }

    static func classImage(playerClass: String) -> NSImage? {
        return imageNamed(playerClass.lowercaseString, from: .Assets)
    }

    static func asset(asset: String) -> NSImage? {
        return imageNamed(asset, from: .Assets)
    }

    static func imageNamed(path: String, from: FromDestination) -> NSImage? {
        switch from {
        case .Bundle:
            let fullPath = NSBundle.mainBundle().resourcePath! + "/Resources/Small/\(path)"
            // DDLogVerbose("Opening image \(fullPath)")
            return NSImage(contentsOfFile: fullPath)

        case .Assets:
            // DDLogVerbose("Opening image \(path)")
            return NSImage(named: path)

        case .Path:
            // DDLogVerbose("Opening image \(path)")
            return NSImage(contentsOfFile: path)
        }
    }
}
